tests/Operation/BulkWriteTest.php

```diff
@@ -192,7 +192,7 @@ public function testReplaceOneReplacementArgumentProhibitsUpdateDocument($replac
     public function testReplaceOneReplacementArgumentProhibitsUpdatePipeline($replacement): void
     {
         $this->expectException(InvalidArgumentException::class);
-        $this->expectExceptionMessage('$operations[0]["replaceOne"][1] is an update pipeline');
+        $this->expectExceptionMessageMatches('#(\$operations\[0\]\["replaceOne"\]\[1\] is an update pipeline)|(\$operations\[0\]\["replaceOne"\]\[1\] to have type "document")#');
```

Comment:
I noted the `replaceOne` validation has three possible exceptions in order:

 * not a document
 * first key is an operator
 * is a pipeline (with `allowEmpty=true`)

Is the "update pipeline" still reachable because `is_document()` simply returns `true` for PHP arrays? What's the data set that triggers "not a document" to be reached here?

Reply:
"Update pipeline" is indeed still reachable because of PHP arrays. Since some of the update pipelines are provided as `PackedArray` instances, these will trigger the "not a document" check before getting to the pipeline checks later on.